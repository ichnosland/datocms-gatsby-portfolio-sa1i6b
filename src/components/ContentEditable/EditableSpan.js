/**
*
* EditableSpan
*
*/
import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';

import media from 'style/mediainjector';
import { EditableFieldStyle } from './index';

const EditableSpan = styled.span`
  display: inline-block;
  min-height: 30px;
  line-height: 0.9;
  vertical-align: middle;
  margin: -1px 6px 0;
  min-width: 30px;
  ${EditableFieldStyle}
  ${({ single }) => single && css`
  display: block;
  margin: 10px auto;
  `}
  ${media.lt667`
    line-height: 1;
  `}
`;

export default EditableSpan;


export class EditableSpanView extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  constructor(...args) {
    super(...args);
    this.node = null;
  }

  componentDidMount = /* istanbul ignore next */ () => {
    const { prefill } = this.props;

    if (this.node && prefill) {
      this.node.innerText = prefill;
    }
  }

  render() {
    const {
      isEditable,
      single,
      spellcheck,
      autocomplete,
      autocorrect,
      autocapitalize,
      onpaste,
      giusto,
      sbagliato,
      onkeydown,
      oninput,
      onfocus,
      onblur,
      oncontextmenu,
      oncopy,
      ondrag,
      ondrop,
      ondragstart,
      ondragend,
    } = this.props;

    return (
      <EditableSpan
        innerRef={/* istanbul ignore next */(node) => { this.node = node; }}
        contentEditable={isEditable}
        single={single}
        spellCheck={spellcheck}
        autoComplete={autocomplete}
        autoCorrect={autocorrect}
        giusto={giusto}
        sbagliato={sbagliato}
        autoCapitalize={autocapitalize}
        onKeyDown={onkeydown}
        onInput={oninput}
        onPaste={onpaste}
        onFocus={onfocus}
        onBlur={onblur}
        onContextMenu={oncontextmenu}
        onCopy={oncopy}
        onDrag={ondrag}
        onDragStart={ondragstart}
        onDragEnd={ondragend}
        onDrop={ondrop}
      />
    );
  }
}

EditableSpanView.propTypes = {
  isEditable: PropTypes.bool,
  single: PropTypes.bool,
  spellcheck: PropTypes.bool,
  autocomplete: PropTypes.string,
  autocorrect: PropTypes.string,
  autocapitalize: PropTypes.string,
  onpaste: PropTypes.func,
  giusto: PropTypes.bool,
  sbagliato: PropTypes.bool,
  onkeydown: PropTypes.func,
  oninput: PropTypes.func,
  onfocus: PropTypes.func,
  onblur: PropTypes.func,
  prefill: PropTypes.string,
  oncontextmenu: PropTypes.func,
  oncopy: PropTypes.func,
  ondrag: PropTypes.func,
  ondragstart: PropTypes.func,
  ondragend: PropTypes.func,
  ondrop: PropTypes.func,
};
