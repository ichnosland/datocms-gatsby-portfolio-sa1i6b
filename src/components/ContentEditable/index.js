/**
*
* ContentEditable
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { transparentize } from 'polished';

import { colore } from 'style/color';

export const EditableFieldStyle = css`
  user-select: text;
  width: ${(props) => props.width};
  max-width: ${(props) => props.maxWidth};
  color: ${(props) => props.color ? props.color : props.theme.brand};
  padding: ${(props) => props.padding ? props.padding : '6px'};
  border-top: 1px solid ${colore.ui.darkBg};
  border-left: 1px solid ${colore.ui.darkBg};
  border-right: 1px solid ${transparentize(0.5, colore.ui.darkBg)};
  border-bottom: 1px solid ${transparentize(0.5, colore.ui.darkBg)};
  border-radius: ${(props) => props.radius ? props.radius : '3px'};
  background-color: ${(props) => props.bgcolor ? props.bgcolor : colore.ui.contrast};
  &:active {
    background-color: ${colore.ui.note};
  }
  &:active,
  &:focus {
    outline: none;
    border-color: ${(props) => props.theme.brand};
  }
  ${({ giusto }) => giusto && css`
    border: 2px solid ${colore.actions.okay};
  `};
  ${({ sbagliato }) => sbagliato && css`
    border: 2px solid ${colore.actions.error};
  `};
`;

export const EditableField = styled.input`
  ${EditableFieldStyle}
  &::placeholder {
    position: relative;
    color: ${(props) => props.theme.subtle};
    opacity: 1;
    overflow: visible;
  }
  &[type="number"] {
    text-align: right;
  }
`;

EditableFieldStyle.defaultProps = {
  theme: {
    brand: '#00BBEF',
    subtle: '#B4E1FF',
  },
};

class ContentEditable extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  render() {
    const {
      content,
      onChangeFunction,
      mustBeNumeric,
      label,
      width,
      giusto,
      sbagliato,
    } = this.props;

    return (
      <EditableField
        onChange={onChangeFunction}
        contentEditable
        placeholder={label}
        value={content}
        type={mustBeNumeric ? 'number' : 'text'}
        width={width}
        giusto={giusto}
        sbagliato={sbagliato}
      />
    );
  }
}

ContentEditable.propTypes = {
  content: PropTypes.string,
  label: PropTypes.string,
  onChangeFunction: PropTypes.func,
  mustBeNumeric: PropTypes.bool,
  giusto: PropTypes.bool,
  sbagliato: PropTypes.bool,
  width: PropTypes.string,
};

export default ContentEditable;
