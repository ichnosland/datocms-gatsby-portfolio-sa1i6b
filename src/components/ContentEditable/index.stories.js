import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text, boolean, select } from '@storybook/addon-knobs';
import { ThemeProvider } from 'styled-components';
import { allThemes, themeOptions } from 'style/theme';
import ContentEditable from './index';
import EditableSpan from './EditableSpan';

storiesOf('ContentEditable', module)
  .addDecorator(withKnobs)
  .add('ContentEditable', () => (
    <ThemeProvider theme={allThemes[select('theme', themeOptions, 'alatin')]}>
      <ContentEditable
        label={text('label', 'Testo placeholder')}
        width={text('width', 'larghezza in px')}
        mustBeNumeric={boolean('mustBeNumeric', false)}
        giusto={boolean('giusto', false)}
        sbagliato={boolean('sbagliato', false)}
        content={text('content (value)', '')}
      />
    </ThemeProvider>
  ))
  .add('EditableSpan', () => (
    <ThemeProvider theme={allThemes[select('theme', themeOptions, 'alatin')]}>
      <p>
        <span>Un po&apos; di testo tanto per gradire</span>
        &nbsp;
        <EditableSpan
          width={text('width', '')}
          maxWidth={text('width', '')}
          single={boolean('single', false)}
          empty={boolean('empty', true)}
          giusto={boolean('giusto', false)}
          sbagliato={boolean('sbagliato', false)}
        />
        &nbsp;
        <span>e vedere che effetto fa su pi√π righe</span>
        &nbsp;
        <EditableSpan
          width={text('width', '')}
          maxWidth={text('width', '')}
          single={boolean('single', false)}
          empty={boolean('empty', true)}
          giusto={boolean('giusto', false)}
          sbagliato={boolean('sbagliato', false)}
        />
        &nbsp;
        <span>come spesso capita.</span>
      </p>
    </ThemeProvider>
  ));

