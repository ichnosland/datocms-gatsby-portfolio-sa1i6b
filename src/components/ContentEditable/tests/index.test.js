import React from 'react';
import { shallow } from 'enzyme';

import ContentEditable, { EditableField } from '../index';

const props = {
  label: 'label',
};

describe('<ContentEditable />', () => {
  it('should render a EditableField', () => {
    const renderedComponent = shallow(
      <ContentEditable
        content="contenuto"
        label="label"
        mustBeNumeric
      />
    );
    expect(renderedComponent.find(EditableField).length).toBe(1);
  });

  it('should render the placeholder if label is provided', () => {
    const renderedComponent = shallow(
      <ContentEditable
        content="contenuto"
        label="label"
        mustBeNumeric
      />
    );
    expect(renderedComponent.find(EditableField).prop('placeholder')).toBe(props.label);
  });

  it('should render input type=number if mustBeNumeric is provided ', () => {
    const renderedComponent = shallow(
      <ContentEditable
        content="contenuto"
        mustBeNumeric
      />
    );
    expect(renderedComponent.find('[type="number"]').length).toBe(1);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render input type=text if mustBeNumeric is not provided ', () => {
    const renderedComponent = shallow(
      <ContentEditable
        content="contenuto"
      />
    );
    expect(renderedComponent.find('[type="text"]').length).toBe(1);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render its css w/ props giusto', () => {
    const renderedComponent = shallow(
      <ContentEditable
        content="contenuto"
        giusto
      />
    );
    expect(renderedComponent.find('[type="text"]').length).toBe(1);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render its css w/ props sbagliato', () => {
    const renderedComponent = shallow(
      <ContentEditable
        content="contenuto"
        sbagliato
      />
    );
    expect(renderedComponent.find('[type="text"]').length).toBe(1);
    expect(renderedComponent).toMatchSnapshot();
  });
});

describe('<EditableField />', () => {
  it('should render a <input> tag', () => {
    const renderedComponent = shallow(
      <EditableField
        content="contenuto"
        label="label"
      />
    );
    expect(renderedComponent.type()).toEqual('input');
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render a its css w/ props color', () => {
    const color = 'red';
    const renderedComponent = shallow(
      <EditableField
        content="contenuto"
        label="label"
        color={color}
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render a its css w/ props padding', () => {
    const padding = '30px';
    const renderedComponent = shallow(
      <EditableField
        content="contenuto"
        label="label"
        padding={padding}
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render a its css w/ props radius', () => {
    const radius = '6px';
    const renderedComponent = shallow(
      <EditableField
        content="contenuto"
        label="label"
        radius={radius}
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render a its css w/ props bgcolor', () => {
    const bgcolor = 'green';
    const renderedComponent = shallow(
      <EditableField
        content="contenuto"
        label="label"
        bgcolor={bgcolor}
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });
});
