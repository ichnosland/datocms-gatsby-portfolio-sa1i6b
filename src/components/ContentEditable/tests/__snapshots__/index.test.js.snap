// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentEditable /> should render input type=number if mustBeNumeric is provided  1`] = `
<ContentEditable__EditableField
  contentEditable={true}
  type="number"
  value="contenuto"
/>
`;

exports[`<ContentEditable /> should render input type=text if mustBeNumeric is not provided  1`] = `
<ContentEditable__EditableField
  contentEditable={true}
  type="text"
  value="contenuto"
/>
`;

exports[`<ContentEditable /> should render its css w/ props giusto 1`] = `
<ContentEditable__EditableField
  contentEditable={true}
  giusto={true}
  type="text"
  value="contenuto"
/>
`;

exports[`<ContentEditable /> should render its css w/ props sbagliato 1`] = `
<ContentEditable__EditableField
  contentEditable={true}
  sbagliato={true}
  type="text"
  value="contenuto"
/>
`;

exports[`<EditableField /> should render a <input> tag 1`] = `
.c0 {
  -webkit-user-select: text;
  -moz-user-select: text;
  -ms-user-select: text;
  user-select: text;
  padding: 6px;
  border-top: 1px solid rgb(209,209,209);
  border-left: 1px solid rgb(209,209,209);
  border-right: 1px solid rgba(209,209,209,0.5);
  border-bottom: 1px solid rgba(209,209,209,0.5);
  border-radius: 3px;
  background-color: rgb(255,255,255);
}

.c0:active {
  background-color: #FFFDDE;
}

.c0:active,
.c0:focus {
  outline: none;
}

.c0::-webkit-input-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0::-moz-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0:-ms-input-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0::placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0[type="number"] {
  text-align: right;
}

<input
  className="c0"
  content="contenuto"
  label="label"
/>
`;

exports[`<EditableField /> should render a its css w/ props bgcolor 1`] = `
.c0 {
  -webkit-user-select: text;
  -moz-user-select: text;
  -ms-user-select: text;
  user-select: text;
  padding: 6px;
  border-top: 1px solid rgb(209,209,209);
  border-left: 1px solid rgb(209,209,209);
  border-right: 1px solid rgba(209,209,209,0.5);
  border-bottom: 1px solid rgba(209,209,209,0.5);
  border-radius: 3px;
  background-color: green;
}

.c0:active {
  background-color: #FFFDDE;
}

.c0:active,
.c0:focus {
  outline: none;
}

.c0::-webkit-input-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0::-moz-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0:-ms-input-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0::placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0[type="number"] {
  text-align: right;
}

<input
  className="c0"
  content="contenuto"
  label="label"
/>
`;

exports[`<EditableField /> should render a its css w/ props color 1`] = `
.c0 {
  -webkit-user-select: text;
  -moz-user-select: text;
  -ms-user-select: text;
  user-select: text;
  color: red;
  padding: 6px;
  border-top: 1px solid rgb(209,209,209);
  border-left: 1px solid rgb(209,209,209);
  border-right: 1px solid rgba(209,209,209,0.5);
  border-bottom: 1px solid rgba(209,209,209,0.5);
  border-radius: 3px;
  background-color: rgb(255,255,255);
}

.c0:active {
  background-color: #FFFDDE;
}

.c0:active,
.c0:focus {
  outline: none;
}

.c0::-webkit-input-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0::-moz-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0:-ms-input-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0::placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0[type="number"] {
  text-align: right;
}

<input
  className="c0"
  color="red"
  content="contenuto"
  label="label"
/>
`;

exports[`<EditableField /> should render a its css w/ props padding 1`] = `
.c0 {
  -webkit-user-select: text;
  -moz-user-select: text;
  -ms-user-select: text;
  user-select: text;
  padding: 30px;
  border-top: 1px solid rgb(209,209,209);
  border-left: 1px solid rgb(209,209,209);
  border-right: 1px solid rgba(209,209,209,0.5);
  border-bottom: 1px solid rgba(209,209,209,0.5);
  border-radius: 3px;
  background-color: rgb(255,255,255);
}

.c0:active {
  background-color: #FFFDDE;
}

.c0:active,
.c0:focus {
  outline: none;
}

.c0::-webkit-input-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0::-moz-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0:-ms-input-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0::placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0[type="number"] {
  text-align: right;
}

<input
  className="c0"
  content="contenuto"
  label="label"
/>
`;

exports[`<EditableField /> should render a its css w/ props radius 1`] = `
.c0 {
  -webkit-user-select: text;
  -moz-user-select: text;
  -ms-user-select: text;
  user-select: text;
  padding: 6px;
  border-top: 1px solid rgb(209,209,209);
  border-left: 1px solid rgb(209,209,209);
  border-right: 1px solid rgba(209,209,209,0.5);
  border-bottom: 1px solid rgba(209,209,209,0.5);
  border-radius: 6px;
  background-color: rgb(255,255,255);
}

.c0:active {
  background-color: #FFFDDE;
}

.c0:active,
.c0:focus {
  outline: none;
}

.c0::-webkit-input-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0::-moz-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0:-ms-input-placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0::placeholder {
  position: relative;
  opacity: 1;
  overflow: visible;
}

.c0[type="number"] {
  text-align: right;
}

<input
  className="c0"
  content="contenuto"
  label="label"
  radius="6px"
/>
`;
